dstep/translator/CommentIndex.d

```diff
@@ -68,6 +68,73 @@ class CommentIndex
         {
             return offset < s ? -1 : (offset == s ? 0 : 1);
         }
+
+        private static bool isNormalized(string content)
+        {
+            import std.ascii : isWhite;
+
+            for (size_t i = 1; i < content.length; ++i)
+            {
+                if (content[i] == '\n' && content[i - 1].isWhite)
+                    return false;
+            }
+
+            return true;
+        }
+
+        private static string normalize(string content)
+        {
+            import std.array : Appender;
+            import std.ascii : isWhite;
+
+            if (isNormalized(content))
+                return content;
+
+            auto result = appender!string;
+
+            size_t lineBegin = 0;
+            size_t firstWhite = 0;
+
+            for (size_t i = 0; i < content.length; ++i)
+            {
+                if (content[i] == '\n')
+                {
+                    if (firstWhite == i)
+                    {
+                        result.put(content[lineBegin .. i + 1]);
+                    }
+                    else
+                    {
+                        result.put(content[lineBegin .. firstWhite]);
+                        result.put('\n');
+                    }
+
+                    lineBegin = i + 1;
+                    firstWhite = i + 1;
+                }
+                else if (!content[i].isWhite)
+                {
+                    firstWhite = i + 1;
+                }
+            }
+
+            if (firstWhite == content.length)
+                result.put(content[lineBegin .. content.length]);
+            else
+                result.put(content[lineBegin .. firstWhite]);
+
+            return result.data;
+        }
+
+        unittest
+        {
+            assert(normalize("") == "");
+            assert(normalize("foo") == "foo");
+            assert(normalize("foo \n") == "foo\n");
+            assert(normalize("foo \n  ") == "foo\n");
+            assert(normalize("foo \n  bar \n  ") == "foo\n  bar\n");
```

Comment:
> It trims extra whitespaces at the end of line. That is after significant source code and before the linefeed character. 

So would this be the same as splitting on newline, stripping trailing whitespace and then joining with newline again? Something like `content.splitter("\n").map!(stripRight).join("\n")`.

> Whitespace before bar are kept because it is bar's indentation.

I see.